### DISCUSSION TITLE
Replace lua with luajit

### DISCUSSION THREAD

**Discussion by xiaobiaozhao** [apache/kvrocks#670] [General]
**Luajit has better performance and I think luajit can be used instead of Lua**

**Luajit official performance report**
https://luajit.org/performance_x86.html

**luajit vs c**
https://mp.weixin.qq.com/s/MDTb9SSkdnOm5VIMEjA9QA

1. If use luajit insteand of lua，kvrocks can run a lua scripte in db side.
2. If lua scripte has high performance，user can move some logic to kvrocks, less network traffic betten db and client.

**Comment by git-hulk** [comment]
@tisonkun proposed this in another thread but the only concern is
that will break the consistency with Redis(Lua 5.1).

**Reply by tisonkun** [reply]
LuaJIT claims that it's compatible with Lua 5.1 but we must have some test coverage to confidently make the change.

**Reply by git-hulk** [reply]
Cool, I didn't notice LuaJIT is compatible with Lua 5.1. If so, I think we can have
a try if @xiaobiaozhao willing to do that. I will be happy to be involved in testing
the compatibility and functionality.

**Comment by PragmaTwice** [comment]
I think this idea is amazing, although some investigation is needed:
- does the API of luajit have a huge difference than lua?
- does luajit have something equal to lua-cjson (or other in-use lua extension)?
- redis did some change to the original lua codebase, could we transplant it to luajit?

**Reply by xiaobiaozhao** [reply]
I have make a kvrocks demo with luajit. It works well.

- luajit and lua5.1 have same api
- lua-cjson works well on luajit
- As far as I know, Redis hasn't changed Lua, just added some Lua extensions that can be ported to Luajit

**Reply by tisonkun** [reply]
@xiaobiaozhao perhaps you can share the branch with your changes or directly send a pull request :)

**Reply by PragmaTwice** [reply]
> I have make a kvrocks demo with luajit. It works well.
> 
> * luajit and lua5.1 have same api
> * lua-cjson works well on luajit
> * As far as I know, Redis hasn't changed Lua, just added some Lua extensions that can be ported to Luajit

You can check here: https://github.com/redis/redis/commits/unstable/deps/lua/src
They did some changes to lua codebases, not only lua extensions (they also modified some logic in lua extensions).
And some issues in lua-cjson with luajit: https://github.com/mpx/lua-cjson/issues/40.

**Reply by PragmaTwice** [reply]
I mean, I'm really looking forward to this change, but this change might be more complicated than expected, so it needs more work.

**Reply by xiaobiaozhao** [reply]
I've already done that

**Reply by PragmaTwice** [reply]
> I've already done that

That is amazing! Look forward to your PR. 🥰

**Reply by xiaobiaozhao** [reply]
Here is my PR with luajit replace lua on my fork.

https://github.com/xiaobiaozhao/incubator-kvrocks/pull/1

**Comment by ShooterIT** [comment]
Some redis issue or PR, FYI
https://github.com/redis/redis/issues/7382
https://github.com/redis/redis/pull/8180

**Comment by ShooterIT** [comment]
+1, go ahead
